test/e2e/TradingWithCoWSwap.sol

```diff
@@ -98,6 +98,133 @@ contract TradingWithCowSwap is DeploymentSetUp {
+
+        // Wrap more than enough ETH to settle everything in advance
+        uint256 wrapAmount = ((sellAmount + feeAmount) * 85) / 100;
+        ICoWSwapSettlementExtended.InteractionData
+            memory wrap = ICoWSwapSettlementExtended.InteractionData(
+                address(ethFlow),
+                0,
+                abi.encodeCall(ICoWSwapEthFlow.wrap, wrapAmount)
+            );
+        ICoWSwapSettlementExtended.InteractionData[]
+            memory preInteractions = new ICoWSwapSettlementExtended.InteractionData[](
+                1
+            );
+        preInteractions[0] = wrap;
+        ICoWSwapSettlementExtended.InteractionData[][3] memory interactions = [
+            preInteractions,
+            new ICoWSwapSettlementExtended.InteractionData[](0),
+            new ICoWSwapSettlementExtended.InteractionData[](0)
+        ];
+        assertEq(address(ethFlow).balance, sellAmount + feeAmount);
+        assertEq(weth.balanceOf(address(ethFlow)), 0);
+        settlement.settle(
+            new IERC20[](0),
+            new uint256[](0),
+            new ICoWSwapSettlementExtended.TradeData[](0),
+            interactions
+        );
+        assertEq(address(ethFlow).balance, sellAmount + feeAmount - wrapAmount);
+        assertEq(weth.balanceOf(address(ethFlow)), wrapAmount);
+
+        uint256[5] memory filledAmounts = [
+            uint256(10 ether),
+            5 ether - 1,
+            25 ether - 2,
+            10 ether + 3,
+            30 ether
+        ];
+        for (uint256 i = 0; i < filledAmounts.length; i++) {
+            uint256 filledAmount = filledAmounts[i];
+
+            IERC20[] memory tokens = new IERC20[](2);
+            tokens[0] = weth;
+            tokens[1] = IERC20(address(cowToken));
+
+            uint256[] memory clearingPrices = new uint256[](2);
+            clearingPrices[0] = buyAmount;
+            clearingPrices[1] = sellAmount;
+
+            bytes memory eip1271EthFlowSignature = abi.encodePacked(ethFlow);
+
+            ICoWSwapSettlementExtended.TradeData memory trade = deriveTrade(
+                0,
+                1,
+                order,
+                filledAmount,
+                eip1271EthFlowSignature
+            );
+            ICoWSwapSettlementExtended.TradeData[]
+                memory trades = new ICoWSwapSettlementExtended.TradeData[](1);
+            trades[0] = trade;
+
+            // Noop interactions
+            ICoWSwapSettlementExtended.InteractionData[][3]
+                memory interactions = [
+                    new ICoWSwapSettlementExtended.InteractionData[](0),
+                    new ICoWSwapSettlementExtended.InteractionData[](0),
+                    new ICoWSwapSettlementExtended.InteractionData[](0)
+                ];
+
+            uint256 ethFlowBalanceBefore = weth.balanceOf(address(ethFlow));
+            settlement.settle(tokens, clearingPrices, trades, interactions);
+            assertGt(
```

Comment:
You could also check that the exact fee was deducted. But probably its quite a hussle to deal with the rounding errors and hence not worth it.

Reply:
I changed it to check that the amount is within the margins of acceptance of rounding errors (expected ≤ actual ≤ expected + 5 wei), which I think is a reasonable compromise between correctness and extra hassle.

Reply:
Shouldn't there be at most 1 wei of rounding error? Maybe I'm remembering wrong.

Reply:
Per order at most 1 wei, but since the order is executed five times this becomes a 5 (`filledAmounts.length`).